dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModelManager.java

```diff
@@ -0,0 +1,669 @@
+                executionModel.addAddressWrite(address, (StoreModel) em);
+            } else {
+                // Should never happen.
+                throw new IllegalArgumentException(String.format(
+                    "Event %s is memory event but neither read nor write", memEvent
+                ));
+            }
+
+        } else if (e.hasTag(Tag.FENCE) && e instanceof GenericVisibleEvent visible) {
+            final String name = (visible).getName();
+            em = new FenceModel(visible, tm, id, name);
+        } else if (e instanceof Assert assrt) {
+            em = new AssertModel(assrt, tm, id);
+        } else if (e instanceof Local local) {
+            em = new LocalModel(local, tm, id);
+        } else if (e instanceof CondJump cj) {
+            em = new CondJumpModel(cj, tm, id);
+        } else {
+            // Should never happen.
+            throw new IllegalArgumentException(String.format("Event %s should not be extracted", e));
+        }
+
+        executionModel.addEvent(e, em);
+    }
+
+    private boolean toExtract(Event e) {
+        // We extract visible events, Locals and Asserts to show them in the witness,
+        // and extract also CondJumps for tracking internal dependencies.
+        return e.hasTag(Tag.VISIBLE)
+               || e instanceof Local
+               || e instanceof Assert
+               || e instanceof CondJump;
+    }
+
+    private void extractMemoryLayout() {
+        for (MemoryObject obj : context.getTask().getProgram().getMemory().getObjects()) {
+            boolean isAllocated = obj.isStaticallyAllocated()
+                                  || isTrue(context.execution(obj.getAllocationSite()));
+            if (isAllocated) {
+                BigInteger address = (BigInteger) evaluateByModel(context.address(obj));
+                BigInteger size = (BigInteger) evaluateByModel(context.size(obj));
+                executionModel.addMemoryObject(obj, new MemoryObjectModel(obj, address, size));
+            }
+        }
+    }
+
+    // Getting the correct relation to extract is tricky.
+    // In the case of redefinition, we care about the one defined last.
+    // If there is no redefinition, we simply return the original one.
+    private Relation getRelationWithName(String name) {
+        // First check if the original definition is asked.
+        if (name.endsWith("#0")) {
+            String originalName = name.substring(0, name.lastIndexOf("#"));
+            return wmm.getRelation(originalName);
+        }
+
+        int maxId = -1;
+        for (Relation r : wmm.getRelations()) {
+            int defIndex = 0;
+            for (String n : r.getNames()) {
+                if (n.startsWith(name + "#")) {
+                    try {
+                        defIndex = Integer.parseInt(n.substring(n.lastIndexOf("#") + 1));
+                        break;
+                    } catch (NumberFormatException e) {}
```

Comment:
If you ignore an exception, you should name the variable `ignored`. Otherwise some IDEs will complain about it.
Also, I think it would be nicer to refactor this "tryParse" part of the method into a utility function like `Optional<Integer> tryParseInteger(String string)`.